src/Simplex/Messaging/Agent/Client.hs

```diff
@@ -67,32 +71,53 @@ newAgentClient cc qSize = do
   sndQ <- newTBQueue qSize
   msgQ <- newTBQueue qSize
   smpClients <- newTVar M.empty
-  subscribed <- newTVar M.empty
+  subscrSrvrs <- newTVar M.empty
+  subscrConns <- newTVar M.empty
   clientId <- (+ 1) <$> readTVar cc
   writeTVar cc clientId
-  return AgentClient {rcvQ, sndQ, msgQ, smpClients, subscribed, clientId}
+  return AgentClient {rcvQ, sndQ, msgQ, smpClients, subscrSrvrs, subscrConns, clientId}
 
 type AgentMonad m = (MonadUnliftIO m, MonadReader Env m, MonadError AgentErrorType m)
 
 getSMPServerClient :: forall m. AgentMonad m => AgentClient -> SMPServer -> m SMPClient
-getSMPServerClient AgentClient {smpClients, msgQ} srv =
+getSMPServerClient c@AgentClient {smpClients, msgQ} srv =
   readTVarIO smpClients
     >>= maybe newSMPClient return . M.lookup srv
   where
     newSMPClient :: m SMPClient
     newSMPClient = do
-      c <- connectClient
+      smp <- connectClient
       logInfo . decodeUtf8 $ "Agent connected to " <> showServer srv
       -- TODO how can agent know client lost the connection?
-      atomically . modifyTVar smpClients $ M.insert srv c
-      return c
+      atomically . modifyTVar smpClients $ M.insert srv smp
+      return smp
 
     connectClient :: m SMPClient
     connectClient = do
       cfg <- asks $ smpCfg . config
-      liftIO (getSMPClient srv cfg msgQ)
+      liftIO (getSMPClient srv cfg msgQ clientDisconnected)
         `E.catch` \(_ :: IOException) -> throwError (BROKER smpErrTCPConnection)
 
+    clientDisconnected :: IO ()
+    clientDisconnected = do
+      removeServerSubs >>= mapM_ (mapM_ removeNotifySub)
```

Comment:
Is it okay that server subscriptions and subscribed connections are removed separately (i.e. not in a single `atomically`)? Maybe it'd be more proper to remove them in a single function returning `Maybe (Set ConnAlias)` and then have a separate function make a notification based on provided set?

Reply:
although don't think it's a huge problem as it is only accessed by a single thread.